File: tags/REL_5_1/Plugins/Kingbotk/AWB Plugin/Article.vb
'Copyright © 2008 Stephen Kennedy (Kingboyk) http://www.sdk-software.com/
'Copyright © 2008 Sam Reed (Reedy) http://www.reedyboy.net/

'This program is free software; you can redistribute it and/or modify it under the terms of Version 2 of the GNU General Public License as published by the Free Software Foundation.

'This program is distributed in the hope that it will be useful,
'but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

'You should have received a copy of the GNU General Public License Version 2 along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

Namespace AutoWikiBrowser.Plugins.Kingbotk
    ''' <summary>
    ''' An object representing an article which may or may not contain the targetted template
    ''' </summary>
    Friend NotInheritable Class Article
        ' Properties:
        Private mArticleText As String, mFullArticleTitle As String
        Private mNamespace As Integer
        Private mEditSummary As String = conWikiPluginBrackets

        ' Plugin-state:
        Private mSkipResults As SkipResults = SkipResults.NotSet
        Private mProcessIt As Boolean ' gets set by ArticleHasAMajorChange/ArticleHasAMinorChange

        ' New:
        Friend Sub New(ByVal ArticleText As String, ByVal vFullArticleTitle As String, _
        ByVal vNamespace As Integer)
            mArticleText = ArticleText
            mFullArticleTitle = vFullArticleTitle
            mNamespace = vNamespace
            'mFullArticleTitle = GetArticleName(mNamespace, mArticleTitle)
        End Sub

        ' Friend properties:
        Friend Property AlteredArticleText() As String
            Get
                Return mArticleText
            End Get
            Set(ByVal value As String)
                mArticleText = value
            End Set
        End Property
        Friend ReadOnly Property FullArticleTitle() As String
            Get
                Return mFullArticleTitle
            End Get
        End Property
        Friend ReadOnly Property [Namespace]() As Integer
            Get
                Return mNamespace
            End Get
        End Property
        Friend Property EditSummary() As String
            Get
                Return mEditSummary
            End Get
            Set(ByVal value As String)
                mEditSummary = value
            End Set
        End Property
        Friend Sub ArticleHasAMinorChange()
            mProcessIt = True
        End Sub
        Friend Sub ArticleHasAMajorChange()
            mProcessIt = True
        End Sub
        Friend ReadOnly Property ProcessIt() As Boolean
            Get
                Return mProcessIt
            End Get
        End Property

        ' For calling by plugin:
        Friend Sub PluginCheckTemplateCall(ByVal TemplateCall As String, ByVal PluginName As String)
            If Not TemplateCall = "" Then ' we have "template:"
                mProcessIt = True
                'EditSummary += "Remove ""template:"", "
                PluginManager.AWBForm.TraceManager.WriteArticleActionLine("Remove ""template:"" call", PluginName, True)
            End If
        End Sub
        Friend Sub PluginIHaveFinished(ByVal Result As SkipResults, ByVal PluginName As String)
            Select Case Result
                Case SkipResults.SkipBadTag
                    mSkipResults = SkipResults.SkipBadTag
                    PluginManager.AWBForm.TraceManager.SkippedArticleBadTag(PluginName, mFullArticleTitle, mNamespace)
                Case SkipResults.SkipRegex
                    If mSkipResults = SkipResults.NotSet Then mSkipResults = SkipResults.SkipRegex
                    PluginManager.AWBForm.TraceManager.SkippedArticle(PluginName, _
                       "Article text matched skip regex")
                Case SkipResults.SkipNoChange
                    PluginManager.AWBForm.TraceManager.SkippedArticle(PluginName, "No change")
                    mSkipResults = SkipResults.SkipNoChange
            End Select
        End Sub

        ' For calling by manager:
        Friend ReadOnly Property PluginManagerGetSkipResults() As SkipResults
            Get
                Return mSkipResults
            End Get
        End Property
        Friend Sub FinaliseEditSummary()
            EditSummary = Regex.Replace(EditSummary, ", $", "")
        End Sub
        Friend Sub PluginManagerEditSummaryTaggingCategory(ByVal CategoryName As String)
            If Not CategoryName = "" Then EditSummary += "Tag [[Category:" + CategoryName + "]]. "
        End Sub

        ' General article writing and manipulation:
        Friend Sub RenamedATemplate(ByVal OldName As String, ByVal NewName As String, ByVal Caller As String)
            DoneReplacement(OldName, NewName, False)
            PluginManager.AWBForm.TraceManager.WriteArticleActionLine( _
               String.Format("Rename template [[Template:{0}|{0}]]→[[Template:{1}|{1}]]", OldName, NewName), Caller)
        End Sub
        Friend Sub DoneReplacement(ByVal Old As String, ByVal Replacement As String, _
        ByVal LogIt As Boolean, Optional ByVal PluginName As String = "")
            mProcessIt = True
            EditSummary += Old + "→" + Replacement + ", "
            If LogIt Then PluginManager.AWBForm.TraceManager.WriteArticleActionLine("Replacement: " + Old + "→" + _
               Replacement, PluginName)
        End Sub
        Friend Sub TemplateAdded(ByVal Template As String, ByVal PluginName As String)
            mEditSummary += String.Format("Added {{{{[[Template:{0}|{0}]]}}}}, ", Template)
            PluginManager.AWBForm.TraceManager.WriteTemplateAdded(Template, PluginName)
            ArticleHasAMajorChange()
        End Sub
        Friend Sub ParameterAdded(ByVal ParamName As String, ByVal ParamValue As String, _
        ByVal PluginName As String, Optional ByVal MinorEdit As Boolean = False)
            mEditSummary += ParamName & "=" & ParamValue & ", "
            PluginManager.AWBForm.TraceManager.WriteArticleActionLine(ParamName & "=" & ParamValue, PluginName)

            If MinorEdit Then ArticleHasAMinorChange() Else ArticleHasAMajorChange()
        End Sub
        Friend Sub RestoreTemplateToPlaceholderSpot(ByVal TemplateHeader As String)
            ' just write one instance of template even if have multiple conTemplatePlaceholder's
            Static strPlaceholder As String = Regex.Escape(conTemplatePlaceholder)
            Static RestoreTemplateToPlaceholderSpotRegex As New Regex(strPlaceholder)

            AlteredArticleText = RestoreTemplateToPlaceholderSpotRegex.Replace(AlteredArticleText, TemplateHeader, 1)
            AlteredArticleText = RestoreTemplateToPlaceholderSpotRegex.Replace(AlteredArticleText, "")
        End Sub
        Friend Sub EditInBrowser()
            Tools.EditArticleInBrowser(FullArticleTitle)
        End Sub

        ' Enum:
        Private Enum BannerShellsEnum
            NotChecked
            NoneFound
            FoundWikiProjectBannerShell
        End Enum

        'State:
        Private WeFoundBannerShells As BannerShellsEnum
        Private MatchEvaluatorString As String

        ' Regexes:
        ' These could probably be simplified significantly (and extra logic doing things like removing linebreaks) if I learnt more of the magic characters
        Private Shared ReadOnly WikiProjectBannerShellRegex As New Regex(conRegexpLeft & WikiProjectBannerShell & _
           ")\b\s*(?<start>\|[^1]*=.*?)*\s*\|\s*1\s*=\s*(?<body>.*}}[^{]*?)\s*(?<end>\|[^{]*)?\s*}}", _
           RegexOptions.Compiled Or RegexOptions.IgnoreCase Or RegexOptions.Singleline Or RegexOptions.ExplicitCapture)


        Friend Shared ReadOnly LineBreakRegex As New Regex("[\n\r]*")
        Private Shared ReadOnly DoubleLineBreakRegex As New Regex("[\n\r]{2,}")

        ' Regex constant strings:
        Private Const WikiProjectBannerShell As String = "WikiProject ?Banner ?Shell|WP?BS|WPBannerShell|WikiProject ?Banners|WPB" ' IGNORE CASE

        ' Match evaluators:
        Private Function WPBSRegexMatchEvaluator(ByVal match As Match) As String
            Const templatename As String = "WikiProjectBannerShell"
            Dim Ending As String = match.Groups("start").Value & match.Groups("end").Value

            ShellTemplateMatchEvaluatorsCommonTasks(templatename, match)

            If Not Ending = "" Then Ending = Microsoft.VisualBasic.vbCrLf + Ending

            Return DoubleLineBreakRegex.Replace("{{" & templatename & "|1=" & Microsoft.VisualBasic.vbCrLf & LineBreakRegex.Replace(MatchEvaluatorString, "") & _
                Microsoft.VisualBasic.vbCrLf & match.Groups("body").Value & Ending & "}}", Microsoft.VisualBasic.vbCrLf)
        End Function

        Private Sub ShellTemplateMatchEvaluatorsCommonTasks(ByVal templatename As String, ByVal match As Match)
            ' Does the shell contain template: ?
            PluginCheckTemplateCall(match.Groups("tl").Value, templatename)
            ' Does the template have it's primary name:
            If Not match.Groups("tlname").Value = templatename Then
                RenamedATemplate(match.Groups("tlname").Value, templatename, templatename)
            End If
        End Sub

        ' Where we (possibly) add our template to an existing shell:
        Friend Sub PrependTemplateOrWriteIntoShell(ByVal Template As Templating, ByVal ParameterBreak As String, _
        ByVal Text As String)
            If WeFoundBannerShells = BannerShellsEnum.NotChecked Then
                If WikiProjectBannerShellRegex.IsMatch(AlteredArticleText) Then
                    WeFoundBannerShells = BannerShellsEnum.FoundWikiProjectBannerShell
                Else
                    WeFoundBannerShells = BannerShellsEnum.NoneFound
                End If
            End If

            Text += Template.ParametersToString(ParameterBreak)

            Select Case WeFoundBannerShells
                Case BannerShellsEnum.FoundWikiProjectBannerShell
                    MatchEvaluatorString = Text

                    AlteredArticleText = WikiProjectBannerShellRegex.Replace(AlteredArticleText, _
                       AddressOf Me.WPBSRegexMatchEvaluator, 1)
                    MatchEvaluatorString = Nothing
                Case BannerShellsEnum.NoneFound
                    AlteredArticleText = Text + AlteredArticleText
            End Select
        End Sub

        ' Misc:
        Friend Function PageContainsShellTemplate() As Boolean
            ' Currently only WPBio can possibly call this, so it's ok to just run the regex and not cache the results
            ' Later on we want to have dynamic redirects and management of these templates (or it maybe should be in
            ' WikiFunctions.TalkPages
            Return WikiProjectBannerShellRegex.IsMatch(AlteredArticleText)
        End Function
    End Class
End Namespace

File: Plugins/Kingbotk Plugin/Article.cs
using AutoWikiBrowser.Plugins.Kingbotk;
using AutoWikiBrowser.Plugins.Kingbotk.Components;
using AutoWikiBrowser.Plugins.Kingbotk.ManualAssessments;
using AutoWikiBrowser.Plugins.Kingbotk.Plugins;

using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Text.RegularExpressions;
using System.Windows.Forms;
using System.Xml;
using WikiFunctions;
using WikiFunctions.Plugin;
//Copyright © 2008 Stephen Kennedy (Kingboyk) http://www.sdk-software.com/
//Copyright © 2008 Sam Reed (Reedy) http://www.reedyboy.net/

//This program is free software; you can redistribute it and/or modify it under the terms of Version 2 of the GNU General Public License as published by the Free Software Foundation.

//This program is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

//You should have received a copy of the GNU General Public License Version 2 along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

namespace AutoWikiBrowser.Plugins.Kingbotk
{
	/// <summary>
	/// An object representing an article which may or may not contain the targetted template
	/// </summary>
	internal sealed class Article
	{
		// Properties:
		private string mArticleText;
		private readonly string mFullArticleTitle;
		private readonly int mNamespace;

		private string mEditSummary = Constants.conWikiPluginBrackets;
		// Plugin-state:
		private SkipResults mSkipResults = SkipResults.NotSet;
			// gets set by ArticleHasAMajorChange/ArticleHasAMinorChange
		private bool mProcessIt;

		// New:
		internal Article(string ArticleText, string vFullArticleTitle, int vNamespace)
		{
			mArticleText = ArticleText;
			mFullArticleTitle = vFullArticleTitle;
			mNamespace = vNamespace;
			//mFullArticleTitle = GetArticleName(mNamespace, mArticleTitle)
		}

		// Friend properties:
		internal string AlteredArticleText {
			get { return mArticleText; }
			set { mArticleText = value; }
		}
		internal string FullArticleTitle {
			get { return mFullArticleTitle; }
		}
		internal int Namespace {
			get { return mNamespace; }
		}
		internal string EditSummary {
			get { return mEditSummary; }
			set { mEditSummary = value; }
		}
		internal void ArticleHasAMinorChange()
		{
			mProcessIt = true;
		}
		internal void ArticleHasAMajorChange()
		{
			mProcessIt = true;
		}
		internal bool ProcessIt {
			get { return mProcessIt; }
		}

		// For calling by plugin:
		internal void PluginCheckTemplateCall(string TemplateCall, string PluginName)
		{
			// we have "template:"
			if (!string.IsNullOrEmpty(TemplateCall)) {
				mProcessIt = true;
			}
		}
		internal void PluginIHaveFinished(SkipResults Result, string PluginName)
		{
			switch (Result) {
				case SkipResults.SkipBadTag:
					mSkipResults = SkipResults.SkipBadTag;
					PluginManager.AWBForm.TraceManager.SkippedArticleBadTag(PluginName, mFullArticleTitle, mNamespace);
					break;
				case SkipResults.SkipRegex:
					if (mSkipResults == SkipResults.NotSet)
						mSkipResults = SkipResults.SkipRegex;
					PluginManager.AWBForm.TraceManager.SkippedArticle(PluginName, "Article text matched skip regex");
					break;
				case SkipResults.SkipNoChange:
					PluginManager.AWBForm.TraceManager.SkippedArticle(PluginName, "No change");
					mSkipResults = SkipResults.SkipNoChange;
					break;
			}
		}

		// For calling by manager:
		internal SkipResults PluginManagerGetSkipResults {
			get { return mSkipResults; }
		}
		internal void FinaliseEditSummary()
		{
			EditSummary = Regex.Replace(EditSummary, ", $", "");
		}

		// General article writing and manipulation:
		internal void RenamedATemplate(string OldName, string NewName, string Caller)
		{
			DoneReplacement(OldName, NewName, false);
		}
		internal void DoneReplacement(string Old, string Replacement, bool LogIt, string PluginName = "")
		{
			mProcessIt = true;
			EditSummary += Old + "→" + Replacement + ", ";
		}
		internal void TemplateAdded(string Template, string PluginName)
		{
			mEditSummary += string.Format("Added {{{{[[Template:{0}|{0}]]}}}}, ", Template);
			ArticleHasAMajorChange();
		}
		internal void ParameterAdded(string ParamName, string ParamValue, string PluginName, bool MinorEdit = false)
		{
			mEditSummary += ParamName + "=" + ParamValue + ", ";

			if (MinorEdit)
				ArticleHasAMinorChange();
			else
				ArticleHasAMajorChange();
		}

        // FIXME: To be replaced
		readonly Microsoft.VisualBasic.CompilerServices.StaticLocalInitFlag static_RestoreTemplateToPlaceholderSpot_strPlaceholder_Init = new Microsoft.VisualBasic.CompilerServices.StaticLocalInitFlag();
		// just write one instance of template even if have multiple conTemplatePlaceholder's
		string static_RestoreTemplateToPlaceholderSpot_strPlaceholder;
		readonly Microsoft.VisualBasic.CompilerServices.StaticLocalInitFlag static_RestoreTemplateToPlaceholderSpot_RestoreTemplateToPlaceholderSpotRegex_Init = new Microsoft.VisualBasic.CompilerServices.StaticLocalInitFlag();
		Regex static_RestoreTemplateToPlaceholderSpot_RestoreTemplateToPlaceholderSpotRegex;
		internal void RestoreTemplateToPlaceholderSpot(string TemplateHeader)
		{
			lock (static_RestoreTemplateToPlaceholderSpot_strPlaceholder_Init) {
				try {
					if (InitStaticVariableHelper(static_RestoreTemplateToPlaceholderSpot_strPlaceholder_Init)) {
                        static_RestoreTemplateToPlaceholderSpot_strPlaceholder = Regex.Escape(Constants.conTemplatePlaceholder);
					}
				} finally {
					static_RestoreTemplateToPlaceholderSpot_strPlaceholder_Init.State = 1;
				}
			}
			lock (static_RestoreTemplateToPlaceholderSpot_RestoreTemplateToPlaceholderSpotRegex_Init) {
				try {
					if (InitStaticVariableHelper(static_RestoreTemplateToPlaceholderSpot_RestoreTemplateToPlaceholderSpotRegex_Init)) {
						static_RestoreTemplateToPlaceholderSpot_RestoreTemplateToPlaceholderSpotRegex = new Regex(static_RestoreTemplateToPlaceholderSpot_strPlaceholder);
					}
				} finally {
					static_RestoreTemplateToPlaceholderSpot_RestoreTemplateToPlaceholderSpotRegex_Init.State = 1;
				}
			}

			AlteredArticleText = static_RestoreTemplateToPlaceholderSpot_RestoreTemplateToPlaceholderSpotRegex.Replace(AlteredArticleText, TemplateHeader, 1);
			AlteredArticleText = static_RestoreTemplateToPlaceholderSpot_RestoreTemplateToPlaceholderSpotRegex.Replace(AlteredArticleText, "");
		}
		internal void EditInBrowser()
		{
			Tools.EditArticleInBrowser(FullArticleTitle);
		}

		// Enum:
		private enum BannerShellsEnum
		{
			NotChecked,
			NoneFound,
			FoundWikiProjectBannerShell
		}

		//State:
		private BannerShellsEnum WeFoundBannerShells;

		private string MatchEvaluatorString;
		// Regexes:
		// These could probably be simplified significantly (and extra logic doing things like removing linebreaks) if I learnt more of the magic characters

        private static readonly Regex WikiProjectBannerShellRegex = new Regex(Constants.conRegexpLeft + WikiProjectBannerShell + ")\\b\\s*(?<start>\\|[^1]*=.*?)*\\s*\\|\\s*1\\s*=\\s*(?<body>.*}}[^{]*?)\\s*(?<end>\\|[^{]*)?\\s*}}", RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.ExplicitCapture);

		static internal readonly Regex LineBreakRegex = new Regex("[\\n\\r]*");

		private static readonly Regex DoubleLineBreakRegex = new Regex("[\\n\\r]{2,}");
		// Regex constant strings:
			// IGNORE CASE
		private const string WikiProjectBannerShell = "WikiProject ?Banner ?Shell|WP?BS|WPBannerShell|WikiProject ?Banners|WPB";

		// Match evaluators:
		private string WPBSRegexMatchEvaluator(Match match)
		{
			const string templatename = "WikiProjectBannerShell";
			string Ending = match.Groups["start"].Value + match.Groups["end"].Value;

			ShellTemplateMatchEvaluatorsCommonTasks(templatename, match);

			if (!string.IsNullOrEmpty(Ending))
				Ending = Environment.NewLine + Ending;

			return DoubleLineBreakRegex.Replace("{{" + templatename + "|1=" + Environment.NewLine + LineBreakRegex.Replace(MatchEvaluatorString, "") + Environment.NewLine + match.Groups["body"].Value + Ending + "}}", Environment.NewLine);
		}

		private void ShellTemplateMatchEvaluatorsCommonTasks(string templatename, Match match)
		{
			// Does the shell contain template: ?
			PluginCheckTemplateCall(match.Groups["tl"].Value, templatename);
			// Does the template have it's primary name:
			if (!(match.Groups["tlname"].Value == templatename)) {
				RenamedATemplate(match.Groups["tlname"].Value, templatename, templatename);
			}
		}

		// Where we (possibly) add our template to an existing shell:
		internal void PrependTemplateOrWriteIntoShell(Templating Template, string ParameterBreak, string Text)
		{
			if (WeFoundBannerShells == BannerShellsEnum.NotChecked) {
				if (WikiProjectBannerShellRegex.IsMatch(AlteredArticleText)) {
					WeFoundBannerShells = BannerShellsEnum.FoundWikiProjectBannerShell;
				} else {
					WeFoundBannerShells = BannerShellsEnum.NoneFound;
				}
			}

			Text += Template.ParametersToString(ParameterBreak);

			switch (WeFoundBannerShells) {
				case BannerShellsEnum.FoundWikiProjectBannerShell:
					MatchEvaluatorString = Text;

					AlteredArticleText = WikiProjectBannerShellRegex.Replace(AlteredArticleText, WPBSRegexMatchEvaluator, 1);
					MatchEvaluatorString = null;
					break;
				case BannerShellsEnum.NoneFound:
					AlteredArticleText = Text + AlteredArticleText;
					break;
			}
		}

		// Misc:
		internal bool PageContainsShellTemplate()
		{
			// Currently only WPBio can possibly call this, so it's ok to just run the regex and not cache the results
			// Later on we want to have dynamic redirects and management of these templates (or it maybe should be in
			// WikiFunctions.TalkPages
			return WikiProjectBannerShellRegex.IsMatch(AlteredArticleText);
		}
		static bool InitStaticVariableHelper(Microsoft.VisualBasic.CompilerServices.StaticLocalInitFlag flag)
		{
			if (flag.State == 0) {
				flag.State = 2;
				return true;
			} else if (flag.State == 2) {
				throw new Microsoft.VisualBasic.CompilerServices.IncompleteInitialization();
			} else {
				return false;
			}
		}
	}
}
